### PR TITLE
libdb: fix c++ compilation with libc++

### PR DESCRIPTION
**Pull request by gjasny**
Specify library name and version:  **libdb/5.3.28**

Otherwise the libdb `atomic_init` define will wreak havoc on the libc++ internals.

Tested with "Conan version 1.59.0":
```
conan create all/conanfile.py libdb/5.3.28@gjasny/testing -pr:b=default -pr:h=default -o libdb:with_cxx=True
```

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
Windows license should have been fixed 🤞

**Comment by prince-chrismc** [comment]
You need to update the imports for this to merge

**Comment by prince-chrismc** [comment]
You can steal my work from #16669 if you like i almost got it working 🤞

**Review by jcar87 — CHANGES_REQUESTED** [review]
Hi @gjasny - thank you for your contribution.

As this recipe is modifying the contents of C++ source files via patches - we are moving towards being a lot more conservative with these moving forward and the Conan Center team will make an assessment before making a decision.

I have a couple of question regarding this PR:
* on what systems/platforms is this an issue? as far as I'm aware, macOS also uses libc++ and  I can see the recipe has binaries for that platform - so is this a Linux only issue?
* what are the actual issues that would otherwise happen without this patch being applied? any evidence in the PR description with a repro example will help us review these type of PRs quicker
* This looks like it could be serious enough to be reported as an issue to the library authors

**Comment by gjasny** [comment]
The reason for the patch is that C11 ships with an "[Atomic operations library](https://en.cppreference.com/w/c/atomic)" whose functions now clash with the ones defined in libdb.

[BerkeleyDB](https://en.wikipedia.org/wiki/Berkeley_DB) has a rough history. It is now owned by Oracle and the license was changed to AGPL. Even if we'd upstream the patch we won't be able to use it (at least not BSD licensed).

Debian (which also packages `v`) also ships two CVE related patches: https://salsa.debian.org/debian/db5.3/-/tree/master/debian/patches

**Comment by gjasny** [comment]
Blocked by #17806

**Comment by jcar87** [comment]
> The reason for the patch is that C11 ships with an "[Atomic operations library](https://en.cppreference.com/w/c/atomic)" whose functions now clash with the ones defined in libdb.

Thanks @gjasny  for added context - the PR description mentions `libc++`, which is the LLVM implementation of the C++ standard library - is this limited to consuming `libdb` when using `libc++` specifically, or _any_ C++ implementation? That is, when the `with_cxx` option is set to `True`.

Please note that for this type of patches we encourage posting error logs and a description of that environment - as part of the review process we make every effort on our end to validate this independently on our end.


> Debian (which also packages `v5.3.28`) also ships two CVE related patches: https://salsa.debian.org/debian/db5.3/-/tree/master/debian/patches

Thank you for providing this link - I can find a patch for fixing identifiers that are prefixed with `__`, but I cannot find a patch to fix the atomic operations? The concern I have, looking at the patch, is that it is renaming identifiers in a header file - if these are functions that are only used privately and there is a documented approach to fixing it this way, we could consider this patch. But otherwise, does it not introduce the risk of changing the names of functions in a public API? That can also break users, which we would like to avoid, so we would need some clarification.

> [BerkeleyDB](https://en.wikipedia.org/wiki/Berkeley_DB) has a rough history. It is now owned by Oracle and the license was changed to AGPL. Even if we'd upstream the patch we won't be able to use it (at least not BSD licensed).

In this context, who is "we"? 
As far as we are concerned in Conan Center, we want to avoid Conan Center becoming a place where forks of libraries are maintained by proxy - our intention is to provide libraries as they are provided by their authors, and patches/fixes should be submitted upstream. We will consider exceptional cases, but if this is an old version that doesn't support C++ contexts with C11,  our first recommendation would be to try out a new version - otherwise Conan Center maintainers become maintainers of a fork of libdb.

**Pull request closed without merging**
